### PR TITLE
Support setting signal masks for child processes on Unix

### PR DESCRIPTION
**Pull request by sunshowers**
This commit implements basic support for setting signal masks during process spawning.
    
* With the `posix_spawn` code path, this means setting `posix_spawnattr_setsigmask`.
* With the `fork/exec` path, this means calling `pthread_sigmask` in the child before executing it.

This PR retains the default behavior (empty signal mask). I've also put up #101077 as a dependent PR to change the default behavior.

API change proposal:

* https://github.com/rust-lang/libs-team/issues/88

Unresolved questions:

* [ ] Is the possible UB with masking `SIGSEGV` etc (documented in the POSIX spec) enough reason to mark this unsafe?

### CODE REVIEW

**Comment by sunshowers** [comment]
Ah I'm not sure if there should be a reviewer at the moment, this is meant to be a draft PR :)

**Comment by jyn514** [comment]
@sunshowers I've unassigned a reviewer for now, but please do run `r? rust-lang/libs-api` once this is ready for review so it doesn't get lost :)

File: library/std/src/sys/unix/process/process_common.rs
Comment (jyn514) [inline library/std/src/sys/unix/process/process_common.rs:172]: Shouldn't this be private?

Comment (sunshowers) [reply library/std/src/sys/unix/process/process_common.rs:172]: It follows the pattern of the other items in the module which are all declared `pub` (though they actually resolve to `pub(crate)`).

File: library/std/src/sys/unix/process/process_common.rs
Comment (joshtriplett) [inline library/std/src/sys/unix/process/process_common.rs:296]: This should accept an `Into<Option<SignalSet>>`, to allow passing `None` as well.

File: library/std/src/sys/unix/process/process_common.rs
Comment (joshtriplett) [inline library/std/src/sys/unix/process/process_common.rs:297]:
```suggestion
    pub fn signal_mask(&mut self, mask: impl Into<Option<SignalSet>>) {
        self.signal_mask = mask.into();
```

Comment (sunshowers) [reply library/std/src/sys/unix/process/process_common.rs]: Thanks, done!

**Comment by sunshowers** [comment]
Updates:
* Addressed review feedback from @joshtriplett
* Added documentation and examples

I think this is ready for review now.

**Comment by sunshowers** [comment]
r? rust-lang/libs-api

**Comment by sunshowers** [comment]
(I still need to write up the change proposal I guess!)

**Comment by thomcc** [comment]
> Is the possible UB with masking SIGSEGV etc (documented in the POSIX spec) enough reason to mark this unsafe?

I guess the problem would be that it changes the documented contract of the `pre_exec` hook. I don't think there's a way to raise `SIGBUS`, `SIGFPE`, or `SIGSEGV` from safe code (well, a way that would violate the UB), but `SIGILL` is also listed as one of the signals that is UB to trigger when it's blocked, and we commonly use that for an "abort immediately" function (e.g. `core::intrinsics::abort()`). (Often this is used for "we need to abort and are inside libcore or liballoc rather than libstd")

That said, `pre_exec` is already unsafe and you have to be careful anyway, so I'm not 100% sure the extent to which this matters. The easy way of triggering sigill from safe code is a double panic, but inside `pre_exec`, the first panic will lead to `sys::abort_internal()`, e.g. `process::abort()` (and not `intrinsics::abort()`), which shouldn't raise SIGILL. (I guess a small caveat here is the `panic_immediate_abort` feature, which does use `intrinsics::abort()`).

I'm not sure if this is worth marking as unsafe, it seems unfortunate if so, but it's probably worth considering carefully if we aren't going to, to avoid a footgun. That said, there's probably enough time to get this right prior to stabilizing, so I don't know that this is a blocker.

**Comment by sunshowers** [comment]
> That said, pre_exec is already unsafe and you have to be careful anyway, so I'm not 100% sure the extent to which this matters.

Yeah, that makes sense. Note that existing code won't be affected though -- only users of both `signal_mask` and `pre_exec` are potentially affected.

**Comment by thomcc** [comment]
Yeah, so I suppose we could adjust the safety contract of `pre_exec` at the same time.

**Comment by sunshowers** [comment]
The other case

**Comment by sunshowers** [comment]
(whoops, closed by accident)

The other issue with safety I was thinking of is: what happens if the child itself has a segfault unrelated to the parent, and the parent has masked `SIGSEGV`? This isn't UB within Rust, but it is UB across the system. I don't know if unsafe extends to that kind of UB.

**Comment by thomcc** [comment]
> I don't know if unsafe extends to that kind of UB.

In general it doesn't. It's totally safe to call processes that crash or do other UB. We have no way of preventing it, it's kind of outside Rust's ability to defend against in general.

That said, fork/exec stuff is ofc where the line gets blurry here (for example, technically `pre_exec` is run inside the child process, but it's still unsafe).

File: library/std/src/sys/unix/process/process_common.rs
Comment (thomcc) [inline library/std/src/sys/unix/process/process_common.rs:63]: Hm, this isn't your fault, but these impls don't match the cited [android impls](https://github.com/aosp-mirror/platform_bionic/blob/ad8dcd6023294b646e5a8288c0ed431b0845da49/libc/include/android/legacy_signal_inlines.h). In particular, it uses `c_long` rather than u8. (`LONG_BIT` is `size_of::<c_long>() * 8`, but we just are using `8`)

I think this will result in incorrect behavior on big endian android, although it's not hard to guess why nobody's noticed before.

Comment (sunshowers) [reply library/std/src/sys/unix/process/process_common.rs:63]: Thanks for catching this! Once #100782 lands I'll rebase and port the changes there over here. (Also I noticed you fixed up the error handling in that PR; thanks for doing that!)

**Comment by zackw** [comment]
What I'm about to describe is a pre-existing bug, but it would be nice either to fix it along with this PR, or else, at least, not make it worse:

It is _incorrect_ to set the signal mask to the empty set in child processes.  This breaks `nohup` and other tools that intentionally spawn a process with signals blocked, expecting that setting to inherit to all processes transitively spawned by that one.  The correct behavior is to _pass down_ the parent process's signal mask, whatever it is, _modified_ by any preferences expressed by the parent.

The `posix_spawnattr_setsigmask` API is expected to be used like this:

```c
sigset_t ss;
pthread_sigmask(SIG_BLOCK, NULL, &ss); // read current thread's signal mask into `ss`
sigaddset(&ss, SIGINT); // for instance
posix_spawnattr_setsigmask(&attr, &ss);
```

Therefore I'd like to ask you to make the following changes:

- Fix the existing bug.  When your new API is not used, the child process's signal mask should default to whatever the parent process's signal mask is, _not_ to empty.  (I was going to say that it should be the parent process's signal mask after undoing any changes made by the Rust runtime on startup, but it looks like the Rust runtime no longer messes with the signal mask, so we're good on that front already.)
- Define your new API in terms of _making changes relative to the parent process's signal mask_, rather than in terms of "here is the brand new signal mask to use."  That could, for instance, look something like this:

    ```rust
    #[unstable(feature = "process_sigmask", issue = "none")]
    fn block_signal(&mut self, sig: std::os::unix::signal::Signal) -> &mut process::Command;
    #[unstable(feature = "process_sigmask", issue = "none")]
    fn unblock_signal(&mut self, sig: std::os::unix::signal::Signal) -> &mut process::Command;
    ```

    where `std::os::unix::signal::Signal` is an enum of all the valid signal constants.

**Comment by sunshowers** [comment]
Thanks for the review, @zackw. I think that makes sense. Note that I will be away for this week and busy next week, so I'll only be able to look at this starting the week of 2022-09-05.

Can the signal to block/unblock just be an i32? The list of signals is pretty long and varies by system, and it seems like letting people use signals from the libc crate would be easiest.

**Comment by zackw** [comment]
> Can the signal to block/unblock just be an i32? The list of signals is pretty long and varies by system, and it seems like having libc be the source of truth and letting people use signals from there would be easiest.

The only problem with that is you have to decide what to do if you are given an argument that's not a valid signal number (`libc::sigaddset` fails).  Because this is a builder, it's not convenient to return a Result.  I _think_ it would be okay to ignore the error, because the most plausible reason for this to happen is if someone's trying to mask a signal that exists on only some systems.  If the signal doesn't exist, then it can't happen, so it doesn't matter if it's masked or not.  But I'd like to know what other people think here.

**Comment by sunshowers** [comment]
> Because this is a builder, it's not convenient to return a Result.

Hmm, is this true? I'm not sure what's generally done in std, but I've definitely written code where consumers have to do:

```rust
MyBuilder::new()
    .fallible_method(arg)
    .expect("arg is valid")
    .fallible_method_2(arg2)?
    .infallible_method()
    .build()
```

**Comment by joshtriplett** [comment]
I agree, the method should return an error.

**Comment by sunshowers** [comment]
Did the following:

* Rebased and incorporated the changes from #100782
* Implemented a new API based on the suggestion from @zackw with some changes (in particular, I added a `will_block_signal` that I think is really useful).
* Updated the API change proposal with this new API
* Split the diff into two commits: the first one to add the new methods to the Android impl, and the second commit to implement signal mask support.

**Comment by sunshowers** [comment]
r? rust-lang/libs-api

**Comment by thomcc** [comment]
There is no libs-api -- it uses the teams from this list (and ignores any prefix like `rust-lang/`): https://github.com/rust-lang/highfive/blob/master/highfive/configs/rust-lang/rust.json#L2-L95. The API review for this will happen on the ACP tho, so I don't mind reviewing the impl, since I think I might have the most review bandwidth at the moment.

r? @thomcc

File: library/std/src/sys/unix/process/process_unix.rs
Comment (thomcc) [inline library/std/src/sys/unix/process/process_unix.rs:331]: This is a behavioral change we should consider separately. If we're going to make it here, then it needs an FCP too.

File: library/std/src/sys/unix/process/process_unix.rs
Comment (thomcc) [inline library/std/src/sys/unix/process/process_unix.rs:524]: ditto.

**Review by thomcc — CHANGES_REQUESTED** [review]
This contains behavioral changes -- the ones suggested in https://github.com/rust-lang/rust/pull/100737#issuecomment-1222672934. So it needs an FCP (which would be a bit confusing since it's an unstable API), or for that to be split into it's own thing.

It's your call, I think the switch to sigemptyset to reading the current signal mask should be split into its own PR (which should be doable, unless I've missed something).

**Comment by sunshowers** [comment]
OK. Well, I was just going by what @zackw said above. I'll back that change out. However, I would like someone else (maybe @zackw, since they suggested the change) to initiate the FCP.

**Comment by sunshowers** [comment]
I've put up #101077 as a dependent PR on top of this one.

File: library/std/src/sys/unix/process/process_unix.rs
Comment (sunshowers) [reply library/std/src/sys/unix/process/process_unix.rs:331]: Moved to #101077.

File: library/std/src/sys/unix/process/process_unix.rs
Comment (sunshowers) [reply library/std/src/sys/unix/process/process_unix.rs:524]: Moved to #101077.

File: library/std/src/sys/unix/process/process_common.rs
Comment (thomcc) [inline library/std/src/sys/unix/process/process_common.rs:401]: We'll want to remove the OnceLock from this if the other PR's FCP doesn't get accepted.

**Review by thomcc — COMMENTED** [review]
The impl looks fine although I'll probably want to review again after the situation with the other PR is resolved.

**Comment by thomcc** [comment]
I'm going to mark this as blocked on the other PR, as it's unlikely that we want this impl without that (and the resolution there may change how we approach this API).

File: library/std/src/os/unix/process.rs
Comment (jsgf) [inline library/std/src/os/unix/process.rs:199]: sigprocmask is probably the more canonical reference.

Also, while the manpage may say that the effect of blocking synchronous traps (ie, SEGV/BUS/FPE/ILL raised while executing an instruction) is undefined, I wouldn't read that as being the same as UB at the Rust/llvm level.

In practice I think it would either deliver the signal as SIG_DFL (ie, Linux), or endlessly re-trap on the same instruction (what x86 macOS seems to do) - either way the program has effectively terminated and won't enter any UB states as far as the program semantics are concerned.

Comment (sunshowers) [reply library/std/src/os/unix/process.rs:199]: Heh it's actually the same man page for both `pthread_sigmask` and `sigprocmask`. I can change the reference in the docs to be `sigprocmask` though.

Comment (jsgf) [reply library/std/src/os/unix/process.rs:199]: Huh - on Fedora they're separate, but pthread_sigmask has almost no content and refers to sigprocmask for discussion of the detailed semantics.

Comment (sunshowers) [reply library/std/src/os/unix/process.rs:199]: (I was referring to the man pages from the official POSIX spec here, as linked in the comment)

**Comment by thomcc** [comment]
I think this is no longer blocked.

**Comment by sunshowers** [comment]
Yep, planning to get to it soon. Meanwhile I've worked around it on beta Rust by calling pthread_sigmask before and after process spawning, which appears to work.

**Comment by JohnCSimon** [comment]
@sunshowers 
ping from triage - can you post your status on this PR? This has sat unmoved for a while now. Thanks!


FYI: when a PR is ready for review, send a message containing 
`@rustbot ready` to switch to `S-waiting-on-review` so the PR is in the reviewer's backlog.

**Comment by Dylan-DPC** [comment]
Closing this due to inactivity. Feel free to make a new pr if you wish to continue this.

**Comment by sunshowers** [comment]
Apologies, I started a new full-time job since I wrote this and haven't had the time to follow up on it. After #101077, I also figured out a way to do this without needing to add new API surface to std:

1. Temporarily use `pthread_sigmask` to set a new sigmask.
2. Spawn the process.
3. Unset the sigmask.

(and use a scope guard to ensure that the sigmask is unset in case of panic.)

This means that I'm probably not going to pursue this further. However, someone else is welcome to.

**Pull request closed without merging**
